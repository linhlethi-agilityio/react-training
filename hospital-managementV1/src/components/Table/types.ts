import { ReactNode } from 'react';

export type TTableAccessor<T> = ((item: T) => ReactNode) | keyof T;

export interface ITableColumn<T> {
  id: string;
  width?: string;
  textAlign?: string;
  accessor: TTableAccessor<T>;
  header?: string;
}

export interface ITableProps<T> {
  className?: string;
  columns: ITableColumn<T>[];
  data: T[];
}
