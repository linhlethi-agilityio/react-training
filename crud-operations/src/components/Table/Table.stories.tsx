import type { Meta } from '@storybook/react';
import { Avatar } from '@chakra-ui/react';

// Components
import Table, { TableColumn } from '.';
import { Student } from '@types';

const meta: Meta<typeof Table> = {
  component: Table,
};

export default meta;

const columnsTable: TableColumn<Student>[] = [
  {
    accessor: () => <Avatar w={65} h={55} borderRadius={8} />,
  },
  {
    header: 'Name',
    accessor: 'name',
  },
  {
    header: 'Email',
    accessor: 'email',
  },
  {
    header: 'Phone',
    accessor: 'phone',
  },
  {
    header: 'Enroll Number',
    accessor: 'enrollNumber',
  },
  {
    header: 'Date of admission',
    accessor: 'dateOfAdmission',
  },
];

const data = [
  {
    enrollNumber: '54547',
    name: 'name 8',
    phone: '7305477760',
    email: 'Carson31@gmail.com',
    dateOfAdmission: 950288400000,
    id: '0',
  },
  {
    enrollNumber: '4545474',
    name: 'A ame 9',
    phone: '7305477760',
    email: 'Ora.Ankunding@yahoo.com',
    dateOfAdmission: 981824400000,
    id: '1',
  },
];

export const Default = () => <Table columns={columnsTable} data={data}></Table>;
