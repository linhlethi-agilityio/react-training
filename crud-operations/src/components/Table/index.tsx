import { ReactNode } from 'react';
import { Table as BaseTable, Tbody, Th, Thead, Tr } from '@chakra-ui/react';

// Components
import TableCell from '@components/TableCell';
import TableRow from '@components/TableRow';

export type TTableAccessor<T> = ((item: T, inputProps?: object) => ReactNode) | keyof T;

export interface ITableColumn<T> {
  id: string;
  fieldName: string;
  accessor: TTableAccessor<T>;
  header: string;
  mWidth?: string;
  width?: string;
}

interface ICustomTableProps<T> {
  columns: ITableColumn<T>[];
  data: T[];
}

const Table = <T,>({ columns, data }: ICustomTableProps<T>) => {
  const headerRow = (
    <Tr>
      {columns.map((columnConfig) => (
        <Th
          key={`header-${columnConfig.id}`}
          textTransform="unset"
          color="tableHeader"
          fontWeight="semiBold"
          lineHeight="xs"
        >
          {columnConfig.header}
        </Th>
      ))}
    </Tr>
  );

  const renderCell = (item: T, accessor: TTableAccessor<T>): ReactNode => {
    if (typeof accessor === 'string') {
      return item[accessor] as ReactNode;
    }

    if (typeof accessor === 'function') {
      return accessor(item);
    }
  };

  return (
    <BaseTable>
      <Thead>{headerRow}</Thead>
      <Tbody>
        {data.map((item, index) => (
          <TableRow key={`table-row-${index}`}>
            {columns.map((columnConfig) => (
              <TableCell key={`table-cell-${index}-${columnConfig.id}`}>
                {renderCell(item, columnConfig.accessor)}
              </TableCell>
            ))}
          </TableRow>
        ))}
      </Tbody>
    </BaseTable>
  );
};

export default Table;
