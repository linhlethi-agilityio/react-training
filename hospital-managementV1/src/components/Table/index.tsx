import { StyledTableCell } from '@components/TableCell';
import { StyledTableHeader } from '@components/TableHeader';
import { StyledTableHeaderCell } from '@components/TableHeaderCell';
import { StyledTableRow } from '@components/TableRow';
import { ReactNode } from 'react';
import { ITableProps, TTableAccessor } from './types';

const Table = <T extends { id: string }>(props: ITableProps<T>) => {
  const { columns, data } = props;

  const renderCell = (item: T, accessor: TTableAccessor<T>): ReactNode => {
    if (typeof accessor === 'string') {
      return item[accessor] as ReactNode;
    }

    if (typeof accessor === 'function') {
      return accessor(item);
    }
  };

  return (
    <table>
      <thead>
        <StyledTableHeader>
          {columns.map((columnConfig) => (
            <StyledTableHeaderCell
              key={columnConfig.id}
              cellWidth={columnConfig.width}
              cellTextAlign={columnConfig.textAlign}
            >
              {columnConfig.header}
            </StyledTableHeaderCell>
          ))}
        </StyledTableHeader>
      </thead>
      <tbody>
        {data.map((item) => (
          <StyledTableRow key={item.id}>
            {columns.map((columnConfig) => (
              <StyledTableCell
                key={`${item.id}-${columnConfig.id}`}
                cellWidth={columnConfig.width}
                cellTextAlign={columnConfig.textAlign}
              >
                {renderCell(item, columnConfig.accessor)}
              </StyledTableCell>
            ))}
          </StyledTableRow>
        ))}
      </tbody>
    </table>
  );
};

export { Table };
