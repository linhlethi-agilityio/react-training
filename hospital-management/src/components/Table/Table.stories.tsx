import { ComponentMeta } from '@storybook/react';
import { Table } from '@components/Table';
import { StyledAvatar } from '@components/Avatar';
import { DownloadIcon } from '@components/Icons';
import { ITableColumn } from './types';

export default {
  title: 'Table',
  component: Table,
} as ComponentMeta<typeof Table>;

interface DataType {
  id: string;
  avatar: string;
  doctor: string;
  date: string;
  dowloadFile?: string;
}

const columns: ITableColumn<DataType>[] = [
  {
    id: '1',
    width: '80px',
    textAlign: 'center',
    accessor: (data: DataType) => <StyledAvatar size="xs" src={data.avatar} />,
  },
  {
    id: '2',
    width: '232px',
    header: 'Doctor',
    accessor: 'doctor',
  },
  {
    id: '3',
    width: '35px',
    accessor: (data: DataType) => (
      <a href={data.dowloadFile}>
        <DownloadIcon size={24} />
      </a>
    ),
  },
];

const data: DataType[] = [
  {
    id: '1',
    avatar: 'https://images.vexels.com/content/145908/preview/male-avatar-maker-2a7919.png',
    doctor: 'Dr. Jacob Jones',
    date: '10/10/2020',
    dowloadFile: '',
  },
  {
    id: '2',
    avatar: 'https://images.vexels.com/content/145908/preview/male-avatar-maker-2a7919.png',
    doctor: 'Dr. Theresa Webb',
    date: '13/10/2020',
  },
];

export const Default = () => <Table columns={columns} data={data}></Table>;
