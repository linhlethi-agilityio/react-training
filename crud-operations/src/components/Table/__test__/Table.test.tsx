import { render } from '@test-utils';

// Components
import Table from '..';

const studentsColumns = [
  {
    header: 'Name',
    accessor: 'name',
  },
  {
    header: 'Email',
    accessor: 'email',
  },
  {
    header: 'Phone',
    accessor: 'phone',
  },
  {
    header: 'Enroll Number',
    accessor: 'enrollNumber',
  },
  {
    header: 'Date of admission',
    accessor: 'dateOfAdmission',
  },
];

const data = [
  {
    enrollNumber: 54547,
    name: 'name 8',
    avatarUrl: 'https://cloudflare-ipfs.com/ipfs/Qmd3W5DuhgHirLHGVixi6V76LhCkZUz6pnFt5AJBiyvHye/avatar/956.jpg',
    phone: '7305477760',
    email: 'Carson31@gmail.com',
    dateOfAdmission: 950288400000,
    id: '0',
  },
];

describe('Table', () => {
  test('should match snapshot for Table', () => {
    const { container } = render(<Table<any> columns={studentsColumns} data={data} />);

    expect(container).toMatchSnapshot();
  });
});
