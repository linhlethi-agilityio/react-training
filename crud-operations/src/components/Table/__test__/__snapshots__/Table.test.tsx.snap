// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table should match snapshot for Table 1`] = `
<div>
  <table
    class="chakra-table css-0"
  >
    <thead
      class="css-0"
    >
      <tr
        class="css-0"
      >
        <th
          class="css-tln5s8"
        >
          Name
        </th>
        <th
          class="css-tln5s8"
        >
          Email
        </th>
        <th
          class="css-tln5s8"
        >
          Phone
        </th>
        <th
          class="css-tln5s8"
        >
          Enroll Number
        </th>
        <th
          class="css-tln5s8"
        >
          Date of admission
        </th>
      </tr>
    </thead>
    <tbody
      class="css-0"
    >
      <tr
        class="css-1w3yp13"
      >
        <td
          class="css-fsw37u"
        >
          name 8
        </td>
        <td
          class="css-fsw37u"
        >
          Carson31@gmail.com
        </td>
        <td
          class="css-fsw37u"
        >
          7305477760
        </td>
        <td
          class="css-fsw37u"
        >
          54547
        </td>
        <td
          class="css-fsw37u"
        >
          950288400000
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
